packages/grpc-js/src/server.ts

```diff
@@ -1421,41 +1420,72 @@ export class Server {
+              this.trace('Connection dropped by keepalive timeout');
+              session.close();
+            }
           }, this.keepaliveTimeoutMs);
-          keepaliveTimeoutTimer.unref?.();
+          keepaliveTimeout.unref?.();
 
           try {
-            session.ping(
-              (err: Error | null, duration: number, payload: Buffer) => {
-                if (keepaliveTimeoutTimer) {
-                  clearTimeout(keepaliveTimeoutTimer);
+            if (
+              !session.ping(
+                (err: Error | null, duration: number, payload: Buffer) => {
+                  clearTimeout(keepaliveTimeout);
+                  if (err) {
+                    if (keepaliveInterval) {
+                      clearInterval(keepaliveInterval);
+                      keepaliveInterval = null;
+                    }
+                    sessionClosedByServer = true;
+                    this.trace(
+                      'Connection dropped due to error with ping frame ' +
+                        err.message +
+                        ' return in ' +
+                        duration
+                    );
+                    session.close();
+                  }
                 }
-
-                if (err) {
-                  sessionClosedByServer = true;
-                  this.trace(
-                    'Connection dropped due to error of a ping frame ' +
-                      err.message +
-                      ' return in ' +
-                      duration
-                  );
-                  session.close();
-                }
-              }
-            );
+              )
+            ) {
+              throw new Error('Server keepalive ping send failed');
+            }
           } catch (e) {
-            clearTimeout(keepaliveTimeoutTimer);
-            // The ping can't be sent because the session is already closed
+            // The ping can't be sent because the session is already closed, max outstanding pings reached, etc
+            clearTimeout(keepaliveTimeout);
+            if (keepaliveInterval) {
+              clearInterval(keepaliveInterval);
+              keepaliveInterval = null;
+            }
+            this.trace(
+              'Connection dropped due to error sending ping frame ' +
+                (e instanceof Error ? e.message : 'unknown error')
+            );
             session.destroy();
           }
         }, this.keepaliveTimeMs);
-        keeapliveTimeTimer.unref?.();
+        keepaliveInterval.unref?.();
       }
 
+      session.once('goaway', (errorCode, lastStreamID, opaqueData) => {
+        if (errorCode === http2.constants.NGHTTP2_ENHANCE_YOUR_CALM) {
+          this.trace('Connection dropped by client due to ENHANCE_YOUR_CALM');
+        } else {
+          this.trace(
+            'Connection dropped by client via GOAWAY with error code ' +
+              errorCode
+          );
+        }
+        sessionClosedByServer = true;
```

Comment:
The log lines indicate that this event means that the session was closed by the client, but this line says that the session was closed by the server. It shouldn't be both.